src/provider/koios.ts

```diff
@@ -0,0 +1,236 @@
+            maxTxExMem: BigInt(result[0].max_tx_ex_mem),
+            maxTxExSteps: BigInt(result[0].max_tx_ex_steps),
+            coinsPerUtxoByte: BigInt(result[0].coins_per_utxo_size),
+            collateralPercentage: parseInt(result[0].collateral_percent),
+            maxCollateralInputs: parseInt(result[0].max_collateral_inputs),
+            costModels: JSON.parse(result[0].cost_models),
+        };
+    }
+
+    async getUtxos(addressOrCredential: Address | Credential): Promise<UTxO[]> {
+        const queryPredicate = (() => {
+            if (typeof addressOrCredential === "string") return addressOrCredential;
+             // should be 'script' (CIP-0005)
+            return addressOrCredential.type === "Key"
+                ? C.Ed25519KeyHash.from_hex(addressOrCredential.hash).to_bech32("addr_vkh")
+                : C.ScriptHash.from_hex(addressOrCredential.hash).to_bech32("addr_vkh");
+        })();
+        try {
+            const result = await this.backendService.getAddressService().getAddressInformation([queryPredicate])
+            if (Array.isArray(result) && result.length > 0 && result[0].utxo_set && result[0].utxo_set.length > 0) {
+                return this.koiosUtxosToUtxos(result[0].utxo_set, result[0].address)
+            } else {
+                return []
+            }
+        } catch (e) {
+            throw new Error("Could not fetch UTxOs from Koios. Try again.");
+        }
+    }
+
+    private async koiosUtxosToUtxos(result: any, address?: string): Promise<UTxO[]> {
+        return (await Promise.all(
+            result.map(async (r: any) => ({
+                txHash: r.tx_hash,
+                outputIndex: r.tx_index,
+                assets: (() => {
+                    const a: Assets = {};
+                    r.asset_list.forEach((am: any) => {
+                        a[am.policy_id + am.asset_name] = BigInt(am.quantity);
+                    });
+                    return a;
+                })(),
+                address: address ? address : r.payment_addr.bech32,
+                datumHash: !r.inline_datum ? r.datum_hash : undefined,
+                datum: r.inline_datum,
+                scriptRef: {
+                    type: r.reference_script ? r.reference_script.type : null,
+                    script: r.reference_script ? applyDoubleCborEncoding(r.reference_script.bytes) : null
+                },
+            })),
+        )) as UTxO[];
+    }
+
+    async getUtxosWithUnit(addressOrCredential: Address | Credential, unit: Unit): Promise<UTxO[]> {
+        const queryPredicate = (() => {
+            if (typeof addressOrCredential === "string") return addressOrCredential;
+            // should be 'script' (CIP-0005)
+            return addressOrCredential.type === "Key"
+                ? C.Ed25519KeyHash.from_hex(addressOrCredential.hash).to_bech32("addr_vkh")
+                : C.ScriptHash.from_hex(addressOrCredential.hash).to_bech32("addr_vkh");
```

Comment:
This was a bug in Blockfrost where you couldn't query bech32 script hashes with the `script` prefix. Is this true for Koios too? Actually it should be `script` instead of `addr_vkh` for script hashes.

Reply:
an issue created in Koios Repo to support Credential Type:

https://github.com/cardano-community/koios-artifacts/issues/187

Reply:
I meant I need examples for possible inputs for addressOrCredential field in Lucid.

Credential type is strange to me
I understand you can have two types of Credentials:
1. Key
2. Script

for Key type you still have 3 more options there:

string, PaymentKeyHash, and StakeKeyHash.

What are the possible values for string?
how is it different than PaymentKeyHash or StakeKeyHash since they are both of type string?

I need some working examples for all types on Mainnet so I would be able to test it against Koios API and discuss with the Koios team to best suit their solution with the requirements.

Reply:
It's either an address or a payment credential (key or script).
- `Address` is of type `string` (bech32), e.g. `"addr..."`
- `Credential` is `{
  type: "Key" | "Script";
  hash: KeyHash | ScriptHash;
};` e.g. `{ type: "Key", hash: "5903e50100..." }`